src/couch_stats/src/csrt_logger.erl

```diff
@@ -0,0 +1,401 @@
+            csrt_server:destroy_resource(PidRef),
+            ok
+    end.
+
+-spec log_process_lifetime_report(PidRef :: pid_ref()) -> ok.
+log_process_lifetime_report(PidRef) ->
+    case csrt_util:is_enabled() of
+        true ->
+            maybe_report("csrt-pid-usage-lifetime", PidRef);
+        false ->
+            ok
+    end.
+
+%% TODO: add Matchers spec
+-spec find_matches(Rctxs :: [rctx()], Matchers :: [any()]) -> matchers().
+find_matches(Rctxs, Matchers) when is_list(Rctxs) andalso is_map(Matchers) ->
+    maps:filter(
+        fun(_Name, {_MSpec, CompMSpec}) ->
+            catch [] =/= ets:match_spec_run(Rctxs, CompMSpec)
+        end,
+        Matchers
+    ).
+
+-spec reload_matchers() -> ok.
+reload_matchers() ->
+    ok = gen_server:call(?MODULE, reload_matchers, infinity).
+
+-spec get_matchers() -> matchers().
+get_matchers() ->
+    persistent_term:get(?MATCHERS_KEY, #{}).
+
+-spec get_matcher(Name :: matcher_name()) -> maybe_matcher().
+get_matcher(Name) ->
+    maps:get(Name, get_matchers(), undefined).
+
+-spec is_match(Rctx :: maybe_rctx()) -> boolean().
+is_match(undefined) ->
+    false;
+is_match(#rctx{}=Rctx) ->
+    is_match(Rctx, get_matchers()).
+
+%% TODO: add Matchers spec
+-spec is_match(Rctx :: maybe_rctx(), Matchers :: [any()]) -> boolean().
+is_match(undefined, _Matchers) ->
+    false;
+is_match(_Rctx, undefined) ->
+    false;
+is_match(#rctx{}=Rctx, Matchers) when is_map(Matchers) ->
+    maps:size(find_matches([Rctx], Matchers)) > 0.
+
+-spec maybe_report(ReportName :: string(), PidRef :: maybe_pid_ref()) -> ok.
+maybe_report(ReportName, PidRef) ->
+    Rctx = csrt_server:get_resource(PidRef),
+    case is_match(Rctx) of
+        true ->
+            do_report(ReportName, Rctx),
+            ok;
+        false ->
+            ok
+    end.
+
+-spec do_lifetime_report(Rctx :: rctx()) -> boolean().
+do_lifetime_report(Rctx) ->
+    do_report("csrt-pid-usage-lifetime", Rctx).
+
+-spec do_status_report(Rctx :: rctx()) -> boolean().
+do_status_report(Rctx) ->
+    do_report("csrt-pid-usage-status", Rctx).
+
+-spec do_report(ReportName :: string(), Rctx :: rctx()) -> boolean().
+do_report(ReportName, #rctx{}=Rctx) ->
+    couch_log:report(ReportName, csrt_util:to_json(Rctx)).
+
+%%
+%% Process lifetime logging api
+%%
+
+-spec get_tracker() -> maybe_pid().
+get_tracker() ->
+    get(?TRACKER_PID).
+
+-spec put_tracker(Pid :: pid()) -> maybe_pid().
+put_tracker(Pid) when is_pid(Pid) ->
+    put(?TRACKER_PID, Pid).
+
+-spec stop_tracker() -> ok.
+stop_tracker() ->
+    stop_tracker(get_tracker()).
+
+-spec stop_tracker(Pid :: maybe_pid()) -> ok.
+stop_tracker(undefined) ->
+    ok;
+stop_tracker(Pid) when is_pid(Pid) ->
+    Pid ! stop,
+    ok.
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+init([]) ->
+    ok = initialize_matchers(),
+    ok = subscribe_changes(),
+    {ok, #st{}}.
+
+handle_call({register, Name, MSpec}, _From, #st{matchers=Matchers}=St) ->
```

Comment:
I don't see where we are sending messages to this clause. If you intend to use it outside of this module add the corresponding access function. I mean something like

```erlang
-module(csrt_logger).

-export([
     register_matcher/2,
 ]).

register_matcher(Name, MSpec) -> 
   gen_server:call(?MODULE, {register, Name, MSpec}, infinity). 
```

Reply:
After thinking about this clause little more I don't think this approach would work. Because we are not using the same ground of true. When we go via `initilize_matchers()` we put them into persistent term. However we do not update the `#st.matchers` used in this clause. Which mean they would go out of sync.

Reply:
Huh odd... looks like the update was hidden for some reason, but I noticed the same thing you did and added that function almost exactly two weeks ago hahaha https://github.com/apache/couchdb/blob/couch-stats-resource-tracker-v3-rebase/src/couch_stats/src/csrt_logger.erl#L113-L116

Good catch on needing to persist registered matchers, I'll add an extra tracking map in the gen_server to merge into the persisted map.